accounts-db/src/file_io.rs

```diff
@@ -0,0 +1,143 @@
+//! File i/o helper functions.
+#[cfg(unix)]
+use std::os::unix::prelude::FileExt;
+use std::{fs::File, ops::Range};
+
+/// `buffer` contains `valid_bytes` of data at its end.
+/// Move those valid bytes to the beginning of `buffer`, then read from `offset` to fill the rest of `buffer`.
+/// Update `offset` for the next read and update `valid_bytes` to specify valid portion of `buffer`.
+/// `valid_len` is # of valid bytes in the file. This may be <= file length.
+pub fn read_more_buffer(
+    file: &File,
+    offset: &mut usize,
+    buffer: &mut [u8],
+    valid_bytes: &mut Range<usize>,
+    valid_len: usize,
+) -> std::io::Result<()> {
+    // copy remainder of `valid_bytes` into beginning of `buffer`.
+    // These bytes were left over from the last read but weren't enough for the next desired contiguous read chunk.
+    buffer.copy_within(valid_bytes.clone(), 0);
```

Comment:
This seems a bit of a footgun? What if a consumer does large reads but short advance_offset()s? This ends up doing large copies.

Don't consumers always know the layout of the file? In which case wouldn't it be better to assert that they always consume everything before the next read so we can avoid this copy?

Reply:
we only copy_within when we want to read more. We only read more when the current buffer doesn't have enough bytes available. Yes, you can use abuse this api. Any better ideas?